### PR TITLE
Update weather API to Tomorrowio

### LINKED ISSUE #535: Weather API is end-of-life
From [this blog post](https://blog.darksky.net/):

> Update: Support for the Dark Sky API service for existing customers will continue until the end of 2022. The iOS app and Dark Sky website will also be available until the end of 2022.

It's not possible to get an API key to test weather features locally, and soon(ish) it will break for the live site too. A replacement needs to be found.

### PR DESCRIPTION
**Pull request by cwille97**
Resolves https://github.com/rubyforgood/Flaredown/issues/535
# Context
The Dark Sky API has been completely shut down for some time now so we are migrating to [Tomorrow.io](https://docs.tomorrow.io/reference/weather-forecast) to bring back weather data for users.

# Approach
I first wrote a [small wrapper](https://github.com/cwille97/tomorrowio_rb) for the API in Ruby and created a [gem](https://rubygems.org/search?query=tomorrowio_rb) for that. This PR removes the old Dark Sky gem and adds my new gem and refactors the weather retriever service to use the new API. By and large the data is mostly backwards compatible with a few exceptions noted below. Notably, it was not necessary to change the Geolocation API since that is still in use. There were some [previous reports](https://github.com/rubyforgood/Flaredown/issues/535#issuecomment-1005453992) of getting the wrong weather data for a given location, so it's possible we may still want to replace the Geolocation API (possibly with the [browser geolocation API](https://developer.mozilla.org/en-US/docs/Web/API/Geolocation_API) if that continues to be a problem.

### Exceptions to backwards compatibility
* This API uses a totally different icon system. I attempted to translate their icon codes into our current icons, however in some cases the best approximation is not very good (i.e. for "thunderstorms" the only option I could pick from is "cloudy"). It would likely improve the user experience to pull in [their icon system](https://docs.tomorrow.io/reference/data-layers-weather-codes) ([assets found here](https://github.com/Tomorrow-IO-API/tomorrow-weather-codes)), while keeping the old icons around for backwards compatibility with existing data.
* Precipitation intensity is calculated slightly differently here. In DarkSky, they just gave one precipitation intensity value whereas Tomorrow gives one for rain, one for sleet, and one for snow. My best guess was to sum these up to make a total precipitation value.
* Tomorrow does not provide a text based summary like Dark Sky did. I made a best approximation of recreating that by combining information from the weather code, as well as the temperature to touch on the core weather attributes for that day in a concise summary. Though I'm open to other suggestions here.

# Tests
Tests to follow, I need to familiarize myself with the VCR package being used here as well as generate the requisite API data for that. Just want to put up an initial draft PR to get some feedback on this so far.

## Other resources
* [A migration guide](https://openweathermap.org/darksky-openweather-3) from a different API, but it translates DarkSky values
<details><summary>Expand for API translations that I came up with</summary>
<p>

| Dark Sky (Forecast API wrapper) | Tomorrow.io | Notes |
|--------|--------|--------|
| Humidity | humidityAvg | |
| Icon | weatherCode | Translating Tomorrow weather codes into the most applicable existing icon |
| precipIntensity | rainIntensityAvg, sleetIntensityAvg, snowIntensityAvg | DarkSky had 1 value for this and my best guess was to sum these 3 |
| pressure | pressureSurfaceLevelAvg | DarkSky uses sea level, Tomorrow uses surface level pressure |
| summary | N/A | I attempted to recreate this value by including information from the weatherCode and the temperature | 
| temperatureMin | temperatureMin | DarkSky defaulted to Kelvin | 
| temperatureMax | temperatureMax | DarkSky defaulted to Kelvin | 

</p>

</details>

### CODE REVIEW

File: backend/Gemfile [hunk 1]
```diff
@@ -48,8 +48,8 @@ gem "globalize"
 # Abort requests that are taking too long
 gem "rack-timeout"
 
-# wrapper for forecast.io API, Dark Sky
-gem "forecast_io"
+# wrapper for tomorrow.io API
+gem "tomorrowio_rb"
```
Comment (compwron) [inline backend/Gemfile]: nice parallelism

File: backend/app/services/weather_retriever.rb [hunk 2]
```diff
@@ -1,6 +1,6 @@
 class WeatherRetriever
   class << self
-    def get(date, postal_code)
+    def get(date, latitude, longitude)
```
Comment (compwron) [inline backend/app/services/weather_retriever.rb]: ncie to not have to go thru postal code for this api

File: backend/app/services/weather_retriever.rb [hunk 3]
```diff
@@ -14,47 +14,82 @@ def get(date, postal_code)
       end
 
       forecast = get_forecast(date, position)
-      the_day = forecast&.daily&.data&.first
 
-      if the_day.blank?
-        Rails.logger.warn "No forecast found for position #{position.inspect}: #{forecast.inspect}"
+      if forecast.status != 200
```
Comment (compwron) [inline backend/app/services/weather_retriever.rb:18]: makes sense 200

File: backend/app/services/weather_retriever.rb [hunk 4]
```diff
@@ -14,47 +14,82 @@ def get(date, postal_code)
...
       )
 
-      Time.zone = tz
-
       forecast
     end
 
-    def create_weather(the_day, position_id)
+    def create_weather(forecast, position_id)
+      today = forecast.body["timelines"]["daily"][0]
```
Comment (compwron) [inline backend/app/services/weather_retriever.rb]: nit: maybe use "dig" for json safety?

File: backend/app/services/weather_retriever.rb [hunk 5]
```diff
@@ -14,47 +14,82 @@ def get(date, postal_code)
...
+        precip_intensity: rain_intensity + sleet_intensity + snow_intensity,
+        pressure: today["pressureSurfaceLevel"],
+        summary: summary,
+        temperature_max: today["temperatureMax"],
+        temperature_min: today["temperatureMin"]
       )
     end
+
+    def get_icon_legacy(today)
+      # Our icons do not coverage their full range of weather codes. We could pull in their icons (linked below) on the frontend to expand options
```
Comment (compwron) [inline backend/app/services/weather_retriever.rb:60]: nice

File: backend/app/services/weather_retriever.rb (hunk 2 above)
Comment (javs-perez) [reply backend/app/services/weather_retriever.rb]: doesn't this break the code below? the method expects `postal_code` in a few places

Comment (cwille97) [reply backend/app/services/weather_retriever.rb]: Whoops, yeah I meant to remove `postal_code` in those places and replace it with coordinates.

File: backend/app/services/weather_retriever.rb (hunk 4 above)
Comment (cwille97) [reply backend/app/services/weather_retriever.rb]: Thanks, I'm pretty unfamiliar with Ruby and did not know about [dig](https://docs.ruby-lang.org/en/3.0/dig_methods_rdoc.html), I will look into that further and make some changes.

**Comment by compwron** [comment]
I am excited for this to go thru!

**Review by compwron — APPROVED** [review]

**Pull request merged**